src/AuthorizedSocketConnection.ts

```diff
@@ -1,49 +1,54 @@
-/* @flow */
-
 import {
   GraphQLError,
+  GraphQLSchema,
+  ValidationContext,
   createSourceEventStream,
   execute,
   parse,
   specifiedRules,
   validate,
 } from 'graphql';
-import type { GraphQLSchema, ValidationContext } from 'graphql';
-import type IoServer from 'socket.io';
+import IoServer, { Socket } from 'socket.io';
 
 import * as AsyncUtils from './AsyncUtils';
-import type { CredentialsManager } from './CredentialsManager';
-import type { CreateLogger, Logger } from './Logger';
-import type { Subscriber } from './Subscriber';
+import { CredentialsManager } from './CredentialsManager';
+import { CreateLogger, Logger } from './Logger';
+import { Subscriber } from './Subscriber';
 import SubscriptionContext from './SubscriptionContext';
 
 export type CreateValidationRules = ({
-  variables: Object,
-  query: string,
-}) => $ReadOnlyArray<(context: ValidationContext) => any>;
-
-type Subscription = {
-  id: string,
-  query: string,
-  variables: Object,
-};
+  variables,
+  query,
+}: {
+  variables: object;
+  query: string;
+}) => ReadonlyArray<(context: ValidationContext) => any>;
+
+interface Subscription {
+  id: string;
+  query: string;
+  variables: object;
+}
 
-type SubscribeOptions<TCredentials> = {
-  hasPermission?: (data: any, credentials: TCredentials) => boolean,
-};
+interface SubscribeOptions<TCredentials> {
+  hasPermission?: (data: any, credentials: TCredentials) => boolean;
+}
+
+interface AuthorizedSocketOptions<TContext, TCredentials> {
+  schema: GraphQLSchema;
+  subscriber: Subscriber<any>;
+  credentialsManager: CredentialsManager<TCredentials>;
+  hasPermission: (data: any, credentials: TCredentials) => boolean;
+  createContext:
+    | ((credentials: TCredentials | null | undefined) => TContext)
+    | null
+    | undefined;
+  maxSubscriptionsPerConnection: number | null | undefined;
+  createValidationRules: CreateValidationRules | null | undefined;
```

Comment:
we probably don't want to allow both `null` and `undefined` here, if we can swing it